### ISSUE TITLE
Don’t show warning about needing h2-h6 in section if have aria-label on section element?

### ISSUE THREAD

**Issue by dstorey** [validator/validator#130]
We get a warning for the following:

Warning: Section lacks heading. Consider using h2-h6 elements to add identifying headings to all sections.
From line 3, column 11543; to line 3, column 11634
````
</section><section role="region" aria-label="site overview"><div c
````

As the aria-label gives a label for the section, reading something like "site overview region" in screen readers that support aria-label, does this remove the need for the warning about h* element? The net result in screen readers is the same. For visual users there is no label/heading but this can often be implied by the layout.

**Comment by sideshowbarker** [comment]
My inclination would be that we should still enjoy the warning, because we didn't add it solely for a11y concerns. But would be good to get @stevefaulkner to chime in on this.

**Comment by sideshowbarker** [comment]
> For visual users there is no label/heading but this can often be implied by the layout.

IMHO that’s usually not the case. Visual users typically navigate through documents based on headings, and `<section>`s that lack headings provide no help to users to navigate through a document unless the `<section>`s have ID anchors and the doc provides some navigation control to jump to each section.

So I would argue that if `<section>s` aren't there for the purpose of helping users navigate through the document, then shouldn't be `<section>s` at all but instead should much better just be `<div>`s. That's literally what `<div>` is for: to mark up division of a document into particular parts for some purpose (but not necessarily for the purpose of providing users direct navigation to those parts, because that's what `<section>` is for.

In other words, there's nothing wrong with just using `<div>`, nor with using nested `<div>s`.

**Comment by dstorey** [comment]
Not sure I agree. Seems weird if you have a series of &lt;section>s that one in the middle of the series suddenly becomes a div instead (along with having zero semantic value passed to assistive tech) just cause it happens to have no visual heading. For example an introduction section that the design calls for not having a &lt;h2>Introduction&lt;/h2> at the top as it is visually obvious it is the introduction, or just a headerless section that is stylistically offset (perhaps with a graphical heading equivalent that aria-label conveys the meaning). It becomes really inconsistent to go to section, section, div, section, section just because there is no visual textual header

**Comment by sideshowbarker** [comment]
> For example an introduction section that the design calls for not having a `<h2>Introduction</h2>` at the top as it is visually obvious it is the introduction, or just a headerless section that is stylistically offset (perhaps with a graphical heading equivalent that aria-label conveys the meaning). It becomes really inconsistent to go to section, section, div, section, section just because there is no visual textual header

The problem is with programmatically detecting cases like that. Of course anything is possible but the core need here is to try to alter authors to cases where they may not be providing adequate means, in the form of headings, for users to navigate among their sections—and cases where maybe they shouldn't be using `<section>` elements at all to begin with.

So I think trying to add further heuristic checking to decrease the possible false positives isn't worth it. The only practical alternative is to never emit the section-lacks-heading warnings at all, and I would really rather not do that because we've gotten feedback from authors saying that it's useful.

Also, the messages are just warnings, not errors, so they can always be safely ignored. And for people using the check Web UI, they can also use the "Message filtering" feature to suppress the messages so that they'll never see them again ever, for any document.

All that said, it would be relatively easy to add an exception for `aria-label`, as you originally proposed. But I'd still like to hear what @stevefaulkner thinks about this, since he's worked on the spec language around it all and knows a lot more than me about what the current behavior is in JAWS, etc., with respect to `<section>` and headings.

**Comment by stevefaulkner** [comment]
As per the [HTML5 spec](http://www.w3.org/TR/html5/sections.html#the-section-element)
"The theme of each section should be identified, typically by including a heading (h1-h6 element) as a child of the section element." 

It encourages use of headings, but does not discourage use of methods such as aria-label in circumstances where a heading does not fit the design in a particular context.

**Comment by sideshowbarker** [comment]
> "The theme of each section should be identified, typically by including a heading (h1-h6 element) as a child of the section element."
> 
> It encourages use of headings, but does not discourage use of methods such as aria-label in circumstances where a heading does not fit the design in a particular context.

OK then, so as far as the checker behavior I'm leaning toward going ahead and implementing @dstorey's suggestion that for a `section[aria-label]`, no "section lacks heading" warning should be emitted.

It makes sense that such sections are likely to have some kind of visual indicator (for example, a border or background in a different color) that already visually distinguishes them as separate sections to sighted users—without needing an explicit heading.

**Comment by stevefaulkner** [comment]
@sideshowbarker sounds good, &lt;section>'s without accessible names are treated like &lt;div>'s. With accessible names they are treated as region's (generic landmarks, with acc name announced in supporting AT/browser combos)
